File: unionfind.py
import sys;
from random import randint;

class UnionFind:

    def __init__(self, n):
        assert(isinstance(n,int) and (n>0));
        self.n = n;
        self.A = [i for i in range(n)];
        self.size = [1 for i in range(n)];
        self.components = {i:0 for i in range(n)};
        
    def find(self, p):
        assert((p >= 0) and (p < self.n));
        while (p != self.A[p]):
            self.A[p] = self.A[self.A[p]];
            p = self.A[p];
        return p;
        
    def union(self, p, q):
        assert((p >= 0) and (p < self.n) and (q >= 0) and (q < self.n));
        if p != q:
            proot = self.find(p);
            qroot = self.find(q);
            
            if (proot != qroot):
                if self.size[proot] <= self.size[qroot]:
                    self.A[proot] = qroot;
                    self.size[qroot] += self.size[proot];
                    del self.components[proot];
                else:
                    self.A[qroot] = proot;
                    self.size[proot] += self.size[qroot];
                    del self.components[qroot];
                    
    def connected(self, p, q):
        assert((p >= 0) and (p < self.n) and (q >= 0) and (q < self.n));
        if p == q: return True;
        proot = self.find(p);
        qroot = self.find(q);
        return (proot == qroot);
        
    def count(self):
        return len(self.components);
        

File: unionfind.cpp
#include <string>
#include <cstring>
#include <cstdlib>
#include "debug.h"
#include "unionfind.h"

#define UFDEFSIZE 512

UNIONFIND::UNIONFIND(int newsize)
{
    A = (UFNode *)0;
    size = 0;
    nElements = 0;
    resize(newsize);
}

UNIONFIND::UNIONFIND(void)
{
    A = (UFNode *)0;
    size = 0;
    nElements = 0;
    resize(0);
}

void UNIONFIND::Init(int n)
{
    if (n < size) resize(n);
    for (int i = 0; i < n; i++)
    {
        A[i].key = i;
        A[i].parent = i;
        A[i].rank = 0;
        A[i].size = 1;
    }
    nElements = n;
}

int UNIONFIND::Find(int k)
{
    ASSERT((k >= 0) && (k < nElements), "UNIONFIND::Find() k out of range.");
    
    if ((k < 0) || (k >= nElements))
        return -1;
    
    int p = k;
    while (p != A[p].parent)
    {
        p = A[p].parent;
    }
    
    return p;
}

void UNIONFIND::Union(int x, int y)
{
    int px = Find(x);
    int py = Find(y);
    
    if (A[px].rank < A[py].rank)
    {
        A[px].parent = py;
        A[py].size += A[px].size;
    }
    else
    {
        A[py].parent = px;
        A[px].size += A[py].size;
        
        if (A[px].rank == A[py].rank)
        {
            A[px].rank++;
        }
    }
}

void UNIONFIND::resize(int newsize)
{
    if (newsize == 0)
    {
        newsize = ((size == 0) ? UFDEFSIZE : size * 2);
    }
    else
    {
        int k;
        for (k = UFDEFSIZE; k < newsize; k *= 2) ;
        newsize = k;
    }
    
    UFNode *p = (UFNode *) malloc(sizeof(UFNode) * newsize);
    memset((void *) p, 0, sizeof(UFNode) * newsize);
    
    if (A) 
    {
        memcpy((void *)p, (void *)A, size * sizeof(UFNode));
        free(A);
    }
    
    A = p;
    size = newsize;
}